### ISSUE TITLE
UX changes

### ISSUE THREAD

**Issue by smichel17** [raatmarien/red-moon#29]
List of changes to consider this ticket closed:

- [ ] Bump to v2.7.0 :smiley:
- [x] Snack bar when top switch is disabled
  - [x] More contrast with dark theme.
- [x]  Remove all drop-down actions in notification except pause/resume.
  - Tapping the text will still open the app.
  - Alternative: replace the 'close' button with "pause for X minutes"
    - There's probably a good way to show this with an icon; I'm not sure what that should look like.
    - When the filter is temporarily paused, the notification text should read "Paused for [X] minutes"
    - If there are 15 or fewer seconds left, "Paused for [X] seconds."
    - Hitting "Pause for X minutes" while the filter is paused will reset the pause time to X minutes.
- [x] Add FAB that toggles pause/resume.
  - [x] The FAB should not be visible when the app bar switch is off
  - [x] Animate the FAB appearing and disappearing like [this](https://www.google.com/design/spec/components/buttons-floating-action-button.html#buttons-floating-action-button-behavior).
- **APP BAR SWITCH**
  - [x] When it's off, all settings **except "Dark theme", "Visit the project page", and "Email me..."** are grayed out and inaccessible.
    - [x] The ADD button should also gray out to reflect that it is not clickable.
    - [x] The three moons should all gray out.
  - [x] When it's off, the notification should not appear.
  - [x] Pausing or resuming the filter via notification or FAB does not change the position.
  - [x] When switched on, the filter should start paused.
    - For this release; the end goal is more nuanced behavior; see #34.
- [x] Add "Location" setting below "turn off time"
  - Grayed out unless "Toggle filter automatically" is set to "sun"
  - Secondary text reads "Lat: ##.## Long: ##.##"
    - If location has never been set, Secondary text reads "Not set"
  - Tapping the setting attempts to update the location.
    - Updates toggle times if location update is successful.
    - [x] If it fails, it shows an error message "Enable location services to update location"  
(Replace the current message starting with "location service isn't enabled...")
- [x] In "sun" mode, secondary text for "Turn filter off" and "Turn filter on" shows in-use (not custom) times.
  - The custom times should still be saved & restored if the user returns to using the custom times.
- [x] Setting "Toggle filter automatically" to "sun" immediately sets toggle times to a reasonable default
  - One suggestion in #5 is approximating sunset/sunrise times [based on time zone](https://github.com/mikereedell/sunrisesunsetlib-java).
  - [x] In "sun" mode, each time the filter is automatically toggled, Red Moon attempts to update location.
    - This attempt fails or succeeds silently.
    - If it succeeds, updates toggle times, too.
    - No change if it fails.
- [x] Remove "Keep running after reboot" setting
  - The service, if enabled, should always stay running after reboot.
- [x] Remove "Contact" header (move these settings under "Other")

---

#### The rest of this comment is out of date. It may still be useful for context.

---

Some of this (anything quoted) is pulling from the Material guidelines ([settings](https://www.google.com/design/spec/patterns/settings.html) and [selection controls](https://www.google.com/design/spec/components/selection-controls.html)). The rest is from my experience as a UX designer (in training).

---

**1. Clarify the how users interact with the app**

_Add a FAB & change notification, the top-level switch, startup settings, & the automatic filter backend._ 

I wrote out a bunch of reasons why the current behavior is confusing, but they're cluttering this section so I moved them down to the bottom of this post <sup>[0](#footnote)</sup> so here we can focus on solutions.

The user doesn't distinguish between the service and the filter. They only care about 3 questions:

- Am I interested in using Red Moon? Probably yes, but this setting should exist so the app can easily be completely disabled without uninstalling it. It's changed so infrequently that it doesn't need to exist outside the app.
  - Do I want Red Moon running right now? Ie, they want to be able to manually stop the service.
    - What parameters should it run under? Both filter parameters and timing. The user configures these (in a profile) and then forgets about it. Maybe occasionally they'll switch profiles. If they're constantly tweaking profile settings, we're doing something wrong or there's a feature we're lacking.

So here's a state hierarchy:

1. Red Moon disabled
2. Red Moon enabled
  1. Filter paused
  2. Filter on
    1. Profile a
    2. Profile b
    3. etc

Each indentation level is a layer of abstraction, and we have controls at each layer. If you mix layers, you'll end up with settings in places that don't make sense<sup>[1](#twilight)</sup>.

The user doesn't care if Red Moon is able to turn off the service in the background, nor if the service starts on boot or if service is able to restart across reboots -- they only care what state the app is in. So the question becomes, where do we control elements of each state, and which states should persist across reboot? Here are my conclusions:

**Remove both 'Startup' settings.** They both have to do with implementation details that users don't care about.
- If a user wants a filter that they have to manually enable and disable that's on after every reboot, they'll already be using a profile where the filter is always on and just pause it manually.

**The top-level switch should enable/disable Red Moon. Its state should persist across reboot.**
  - The only way to toggle this setting is tapping the slider in the app.
    - control via intent can happen at a higher level
    - **Remove the 'close' button from the notification.** It would be too easy for a user to disable Red Moon when they intended to turn off the filter. If you want to keep this button, repurpose it. See the 2nd comment in this thread.

**Add a FAB that pauses and resumes the filter. Its state should _not_ persist across reboot.**
- If Red Moon is disabled, the FAB will enable it as well as resuming the filter.
- The 'pause' button in the drop-down notification will have identical behavior, except it'll never be shown while Red Moon is disabled.
- If you use this button while Red Moon is disabled (state 1), it will enable red moon and turn the filter on.

Changes to the filter settings get their own section at the bottom.

---

**2. Change "Dark theme" to a switch and remove the secondary text.**

>Secondary text is optional. If the label is sufficient on its own, don't add a secondary text description.
[...]
>If you have a single option, avoid using a checkbox and use an on/off switch instead.

"Dark theme" is pretty universally understood.

Switches are pretty overrated; they're often more confusing than a checkbox would be. That said, they have their place on lone settings or when changing the setting has an immediate effect (like toggling wifi or switching theme). Dark theme is both.

---

**3. Shuffle around settings under the 'automatic filter' header**

>For non-switch settings, secondary text should show the current status of a setting only. 

The first 3 lines mimic Twilight's interface. 'Location' should be smaller than Twilight's and only show the coordinates in its secondary text.

~~Automatic filter~~ Filter times
- {Always | Sun | Custom}  
_[slider bar for viewing current times or setting them if 'custom' is selected]_
- Location  
_[coordinates]_

'Location' opens up a new activity that guides you through the process of picking your location. That screen has 4 options:

- Automatic (accurate)  
_Requires location services_
- Automatic (approximate)  
_Estimate location from system date/time_
- Enter coordinates
- Search by city

---

<a name="footnote">0</a>: It should be possible to understand how all settings will affect behavior without actually testing out different combinations of settings. It boils down to whether there a difference between a filter that's on but paused and a disabled filter. Ie, can the service be running and the filter be disabled? Related questions that follow:

- Does the notification showing mean the service is running or that the filter is on?
- If I have I have automatic filter times set but I pause the filter, will the filter turn on at sunset or be paused indefinitely?
- Does the top level switch start and stop the filter or the service?
- Does the 'automatic filter' setting start and stop the filter or the service?
- Does the stop button in the notification stop the service or filter?

My best guesses for the answers are:

- The notification tells whether the service is running or not, as does the switch.
- The filter turns on when the service starts, but can be paused through the notification.
- Automatic filter toggles the service on and off.

Could you explain what the actual relationship between filter and service is? If it's different enough from my guess that implementing my suggestions would be difficult, I'll make some modifications.

<a name="twilight">1</a>: It's pretty clear that the team behind Twilight did their homework for the design, but they got lost in the details. They decided "always on" should be at the manual-intervention level; I think it's a profile setting. They also decided "filter times" should be a global setting, when I think it's also a profile setting. Changing it is larger-scope than this ticket, though. I am only including changes here that I think should be uncontroversial.

**Comment by smichel17** [comment]
If we keep extending the state hierarchy:

1. Red Moon disabled
2. Red Moon enabled
  1. Filter on
  2. Filter paused
    1. Profile
      1. Filter auto-on
      2. Filter auto-off

With this, we can re-add the 'close' button in the notification that we "removed" (if these weren't presented together) above. Its function is now to set the program state to 2.ii.a.b and close the notification. That means the filter will be disabled until the next 'auto-on' time.

You could add a global option "Always show notification." If this setting is disabled, the notification doesn't show when Red Moon is in state 2.ii.a.b, and if restarting the service at a given time is possible, you can kill the service during that time.

**Comment by smichel17** [comment]
(3) in the original post is related to #5

**Comment by raatmarien** [comment]
Thanks a lot for these insights! I don't have much design experience myself, so this really helps a lot.

I will go over this in detail after I've relicensed Red Moon to the GPLv3+.

**Comment by raatmarien** [comment]
I've started with the easiest part, I've changed the CheckBoxPreference for a dark theme to a SwitchPreference and removed the summary. I hope this satisfies point 2.

**Comment by smichel17** [comment]
I tried the most recent changes. I think they're a great improvement! Actually, enough of an improvement (and time to think) that I've changed my mind about some features. I also think this thread is getting a bit hard to follow (and want to keep the scope reasonable), so I'm going to give a list of features to close this issue and omit the context\*.

- [ ]  Remove all drop-down actions in notification except pause/resume.
  - Tapping the text will still open the app.
- [ ] Add FAB that toggles pause/resume.
- [ ] Top-level switch enables/disables the notification _& the FAB_
  - Pausing the filter via notification or FAB does not turn the switch off
- [ ] Add "Location" setting below "turn off time"
  - Grayed out unless "Toggle filter automatically" is set to "sun"
  - Secondary text reads "Lat: ##.## Long: ##.##"
    - If location has never been set, Secondary text reads "Not set"
  - Tapping the setting attempts to update the location.
    - Updates toggle times if location update is successful.
    - [ ] If it fails, it shows an error message "Enable location services to update location" (instead of the current message starting with "location service isn't enabled...")
- [ ] In "sun" mode, secondary text for "Turn filter off" and "Turn filter on" shows the times in use (not the custom times)
  - The custom times can still be saved & used if the user switches back to using the custom times.
- [ ] Setting "Toggle filter automatically" to "sun" immediately sets toggle times based on time zone.
  - [ ] When in "sun" mode, every time the filter is automatically toggled, Red Moon attempts to update location.
    - This attempt fails or succeeds silently.
    - If it succeeds, updates toggle times, too.
    - No change if it fails.
- [ ] Remove "Keep running after reboot" setting
- [ ] Remove "Contact" header (move these settings under "Other")

\*If you disagree with any of these or have questions, ask and I'll give my reasoning, I just don't want to clutter the comment if everyone agrees with the suggestions.

**Comment by raatmarien** [comment]
Thanks a lot, this is very helpful!

>- [ ]  Remove all drop-down actions in notification except pause/resume.

I'm wondering what your reasoning is for not including a stop button in the notification. Does a stop button in the notification make it too easy for the user to accidentally stop the service?

I'm also thinking of adding a "pause for 30 min" button to the notification and was wondering how you would feel about that? Would it be helpful to the user, or overly complicated?

>- [ ] Remove "Keep running after reboot" setting

Without this setting, I think the filter state should persist after a reboot. Would you agree?

>- [ ] Add FAB that toggles pause/resume.

Should the FAB just be the primary colour of Red Moon (red), or should I choose a tint colour?

Thanks again!

**Comment by smichel17** [comment]
>I'm wondering what your reasoning is for not including a stop button in the notification. Does a stop button in the notification make it too easy for the user to accidentally stop the service?

Exactly :smile: 

I think the top-level switch should be the only place a user can enable/disable the service. If there's some way to stop the service from running when the filter is off and then automatically start it again at a specified time that might be useful, too -- but that's an implementation detail to save battery, not something visible to the user, who's thinking in terms of "Red Moon is enabled/disabled".

>I'm also thinking of adding a "pause for 30 min" button to the notification and was wondering how you would feel about that? Would it be helpful to the user, or overly complicated?

I think this is a good idea, but I'm also trying to limit the scope of this issue so it's easier to follow. However, if it's easier to do that than switch to a 1-button notification, by all means go for it :)

>Without this setting, I think the filter state should persist after a reboot. Would you agree?

Yes!

>Should the FAB just be the primary colour of Red Moon (red), or should I choose a tint colour?

This is a visual design question, which I don't have any particular expertise in. I skimmed the Material design guidelines on [color](https://www.google.com/design/spec/style/color.html) and [FABs](https://www.google.com/design/spec/components/buttons-floating-action-button.html) and it seems like it makes sense to use the primary color; it might also make sense to use the small FAB variant, but that'd be an easier decision to make if I could see it in action (ie, just pick one and go with it and we can assess if it's better to use small or normal-sized later).

**Comment by smichel17** [comment]
Do you get notifications when I edit my comments, or should I also make a comment here whenever I edit the list?

>- [ ] Setting "Toggle filter automatically" to "sun" immediately sets toggle times to a reasonable default
>  - One suggestion in #5 is approximating sunset/sunrise times based on time zone.

**Comment by raatmarien** [comment]
Thanks for the answers!

>Do you get notifications when I edit my comments, or should I also make comment saying what I edited?

I don't get any notifications when you edit your comments, so a comment would be helpful.

>- [ ] Setting "Toggle filter automatically" to "sun" immediately sets toggle times to a reasonable default
>  - One suggestion in #5 is approximating sunset/sunrise times based on time zone.

Sadly, the timezone doesn't provide any information for the sunset and sunrise times in the users local times. However, I think it is a good idea to provide a default for sunrise and sunset times instead of using the custom times.

**Comment by smichel17** [comment]
>Sadly, the timezone doesn't provide any information for the sunset and sunrise times in the users local times. However, I think it is a good idea to provide a default for sunrise and sunset times instead of using the custom times.

Isn't that what [sunrisesunsetlib-java](https://github.com/mikereedell/sunrisesunsetlib-java) could do?

**Comment by raatmarien** [comment]
>Isn't that something sunrisesunsetlib-java could do? 

No, unfortunately sunsetsunriselib-java needs a location and a timezone in which to return the time. See https://github.com/mikereedell/sunrisesunsetlib-java/blob/master/src/main/java/com/luckycatlabs/sunrisesunset/SunriseSunsetCalculator.java.

**Comment by smichel17** [comment]
This addresses the first part of your comment on #33, and also reflects my latest thinking about how states should work. Essentially, there should only be on/off. You can manually toggle the filter on/off; red moon also includes settings to automatically turn the filter on and off at certain times. The app does not distinguish between a manual or automatic start/stop.

This makes both the code base and the user interface way simpler.

>- [ ] Remove all references to "Pausing" or "resuming" the filter. Refer to these as "turning the filter off/on"

**Comment by raatmarien** [comment]
Should the filter automatically turn on at the set time if the user has disabled Red Moon?

**Comment by smichel17** [comment]
>Should the filter automatically turn on at the set time if the user has disabled Red Moon?

No. If Red Moon is disabled, the Red Moon service should be off, there should be no notification, etc.

I have used the language "pause/resume" across this thread to be consistent about what functionality I'm talking about and it's confusing to have a todo in here that talks about changing that vocab, so I moved that part to its own issue, #35.

**Comment by smichel17** [comment]
Added:

>  - [ ] When the [top level] switch is off, all other settings are grayed out and inaccessible.

**Comment by smichel17** [comment]
Added:

>  - [ ] When switched on for the first time, the filter should start paused.
>    - This is how things should be for this release; the end goal is more nuanced behavior; see #34.

**Comment by raatmarien** [comment]
> No. If Red Moon is disabled, the Red Moon service should be off, there should be no notification, etc.

Great, thanks for the quick answer.

>  - [ ] When the top-level switch is off, all other settings are grayed out and inaccessible.

Could you elaborate your reasoning on this? I'm afraid users might be confused as to why they are unable to edit any settings.

>  - [ ] When switched on for the first time, the filter should start paused.

With switched on, do you mean when the user first opens the app or when the user first presses the top switch?

**Comment by breversa** [comment]
(What's a FAB, anyway ?)

**Comment by smichel17** [comment]
@breversa [Floating Action Button](https://www.google.com/design/spec/components/buttons-floating-action-button.html)

**Comment by smichel17** [comment]
>Could you elaborate your reasoning on this? I'm afraid users might be confused as to why they are unable to edit any settings.

The reason is, it provides a VERY strong visual queue that this switch enables and disables ALL functionality. It's the same reason we gray out the "Location" setting when using custom times. Sure, we could technically set the location, but the setting isn't in use, so we gray it out.

The top switch is always visible. If it's the only toggle-able thing, I think it should be clear. Particularly, after the user has toggled it ONCE, seeing all the settings suddenly become un-gray will be *such* strong feedback that they will never be confused again.

It is a good point, though, that the *first* time a user opens the app (or if a different user is borrowing the phone and opens the app) they may be confused. Added a few lines to the top post.

>With switched on, do you mean when the user first opens the app or when the user first presses the top switch?

I mean any time the top switch moves from 'off' to 'on'.

As noted, this is not the long-term plan (if you set automatic filter and then disable Red Moon, then enable it again, the filter should be off if it's night and on if it's day). However, that functionality is #34, and I'd rather not hold up the next release for it :)

**Comment by raatmarien** [comment]
>  - [ ] When it's off, the notification _& the FAB_ should both disappear.

Is it important that the FAB disappears? I would think it provides an extra option for users who only want to see the notification if the filter is actually on and want a way to instantly enable Red Moon and turn on the filter manually.

>  - [ ] When the app is opened, if the switch is off, there is a visual indicator prompting the user to turn it on.
    - I'm imagining some text saying "Enable Red Moon" and an (animated -- moves up and down a bit?) up-arrow in the upper right corner pointing to the switch.
    - [ ] When the user scrolls through the (disabled) settings, this indicator should disappear.

I will look into how to do this. I'm not sure it will be easy to have an arrow pointing to the switch, since it can be in different places depending on the device (think RTL layouts).

**Comment by smichel17** [comment]
>Is it important that the FAB disappears? I would think it provides an extra option for users who only want to see the notification if the filter is actually on and want a way to instantly enable Red Moon and turn on the filter manually.

At the time I was thinking that the user should have to turn on the top-level switch to enable the app, period -- to force the user to understand the function of the top toggle (vs the FAB or notification).

>I will look into how to do this. I'm not sure it will be easy to have an arrow pointing to the switch, since it can be in different places depending on the device (think RTL layouts).

It doesn't need to be an arrow\*. For example, you could dim the whole screen except for a circle around the toggle, and maybe add some descriptive text, and that would do the job just as well.

---

That said... I do still think it would be better to have the FAB disappear and a visual queue that guides you to turning on the top-level switch. However, if we kept the FAB, I wouldn't be worried about users finding a way to turn on the app anymore, so the visual queue is no longer needed.

That is to say, I think it's a nice-to-have and outside the scope of this ticket. Almost done! :smiley: 

\*The purpose i to highlight "this is the action you are expected to take." As to what it should look like -- honestly that's a visual design thing that I don't have any special knowledge or strong feelings about.

**Comment by smichel17** [comment]
Just tested this out, and I have a few thoughts:

>- [ ] The ADD button should also gray out to reflect that it is not clickable.
>- [ ] The three moons should all gray out.
>- [ ] The three 'other' settings do not need to be grayed out.

Mockup of the faded stuff:

![gray moons and ADD button](https://cloud.githubusercontent.com/assets/6563664/14336327/9892e350-fc30-11e5-87bd-8857483f09e0.png)

After trying it, I do think it is confusing at first -- even with the FAB -- why all the settings are disabled. As I expected, after toggling the app bar switch (I might as well call it the right name) once, it was immediately clear what's going on. The FAB actually makes this worse: when you use the FAB to enable, there's so much activity that it's easy to miss the app bar switch turning on, which means the user may not learn what it does. So, I'll go back to what I said at first. See also: [the animation](https://www.google.com/design/spec/components/buttons-floating-action-button.html#buttons-floating-action-button-behavior)

> - [ ] The FAB should not be visible when the app bar switch is off

Just removing the FAB right away doesn't solve things -- it's still too difficult to notice the app bar slider. I made a super fast & lazy mockup, that I think does the job:

![disabled](https://cloud.githubusercontent.com/assets/6563664/14336338/c1f47394-fc30-11e5-8110-d5ede498758b.png)

Implementation details:

- The mockup is black with 75% opacity. This may or may not look best when actually implemented.
- [ ] It should be at the same level as the app bar (4dp high, I believe).
  - This puts it under the FAB (6dp), so the FAB can animate on to the screen while the overlay is sliding off.
- I am not certain exactly what animations should be used. My best guesses:
  - [ ] The overlay should be in place when the app loads.
  - [ ] Upon turning on the app bar switch, the overlay should slide downward off the screen without changing elevation.

**Comment by raatmarien** [comment]
When Red Moon is disabled, the FAB fades out and a snackbar with help fades in. I chose this over a black overlay, because some settings remain accessible when Red Moon is disabled.

Screenshot:

<img src="https://lut.im/uBvJvDQw9j/sGkCugnib53mgXkr.png" width="360" height="640" />

>- [ ] Bump to v3.0 :smiley:
  - Removed settings -> broke backwards compatibility -> major version bump!
  - Or maybe I'm looking at things wrong and it should just be 2.7.0.

I feel v2.7.0 is more appropriate, since users can upgrade without loss of settings. 

If you agree it is ready, I will release the new version and close this issue.

**Comment by smichel17** [comment]
>I feel v2.7.0 is more appropriate, since users can upgrade without loss of settings.

No strong feelings, sounds good!

>If you agree it is ready, I will release the new version and close this issue.

One last thing: the snack bar needs a little more contrast with the background when the dark theme is enabled. Options I can think of: Lighten it a bit, darken it further, add a lighter gray line along the top.

![snackbar-dark](https://cloud.githubusercontent.com/assets/6563664/14389040/c35a510a-fd7e-11e5-8178-563cc760c78f.png)

**Comment by smichel17** [comment]
>When Red Moon is disabled, the FAB fades out and a snackbar with help fades in. I chose this over a black overlay, because some settings remain accessible when Red Moon is disabled.

I was thinking black overlay only on first run (once the user has toggled once, they'll remember how to replicate the behavior in the future), but I like this solution more.

**Comment by raatmarien** [comment]
> One last thing: the snack bar needs a little more contrast with the background when the dark theme is enabled. Options I can think of: Lighten it a bit, darken it further, add a lighter gray line along the top.

Nice catch, I changed it to the standard android white when dark mode is enabled for contrast.

**Comment by raatmarien** [comment]
[Version 2.7.0 released!](https://github.com/raatmarien/red-moon/releases/tag/v2.7.0)

Thank you so much for all your hard work! The UX a lot better now :smile: :+1:.

**Comment by M2ck** [comment]
@raatmarien just a small detail, you should replace the checkbox by a switch, Material Design specs say about checkboxes:
>Checkboxes allow the user to select multiple options from a set.
If you have multiple options appearing in a list, you can preserve space by using checkboxes instead of on/off switches.
**If you have a single option, avoid using a checkbox and use an on/off switch instead.**

And about switches:
>**On/off switches toggle the state of a single settings option.**
The option that the switch controls, as well as the state it’s in, should be made clear from the corresponding inline label. Switches take on the same visual properties of the radio button.
The on/off slide toggle with the text “on” and “off” included within the asset is deprecated. Use the switch shown here instead.

Source: [Components – Selection controls](https://www.google.com/design/spec/components/selection-controls.html#)

**Comment by smichel17** [comment]
@M2ck I made a new issue for this, #49

**Comment by M2ck** [comment]
@smichel17 Thanks!
